File: Array/399.Evaluate_Division/Python/399.Evaluate_Division.py
class Solution(object):
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """
        #==================================#
        # DFS & BFS based traversal method #
        #==================================#

        ############
        #Initialize
        ##### Length of equations, queries #####
        len_equal = len(equations)
        len_queries = len(queries)

        ##### Record, Traversal dictionary/hashmap #####
        record_dict = {}
        record_trav_dict = {}

        ##### Result array #####
        res_arry = []


        #################################################################
        #DFS & BFS based loop traversal with recorded dictionary/hashmap
        ##### Step 1: Record indexed-elements with dictionary/hashmap #####
        for equal_idx in range(len_equal):
            record_equation = equations[equal_idx] #Record equation

            ##### Check if the current indexed-element existed or not #####
            if (record_equation[0] not in record_dict):
                record_dict[record_equation[0]] = [[record_equation[1], values[equal_idx]]] #Keep updating/recording
            else:
                (record_dict[record_equation[0]]).append([record_equation[1], values[equal_idx]]) #Keep updating/recording

            if (record_equation[1] not in record_dict):
                record_dict[record_equation[1]] = [[record_equation[0], (1. / values[equal_idx])]] #Keep updating/recording
            else:
                (record_dict[record_equation[1]]).append([record_equation[0], (1. / values[equal_idx])]) #Keep updating/recording

        ##### Step 2: Looped-traversal with recorded dictionary #####
        for queries_idx in range(len_queries):
            record_query = queries[queries_idx] #Record query

            res_arry.append(self.dfsbfsTraversal(record_dict, record_query[0], record_query[1], 1., record_trav_dict)) #Recursion function call

            record_trav_dict.clear() #Reset

        return res_arry


    def dfsbfsTraversal(self, record_dict, start_query, end_query, res_scl, record_trav_dict):
        """
        :type record_dict: dict
        :type start_query: char
        :type end_query: char
        :type res_scl: float
        :type record_trav_dict: dict
        :rtype: float
        """
        #====================================#
        # DFS & BFS based traversal function #
        #====================================#
        if ((start_query not in record_dict) or (end_query not in record_dict)): #Issue/Boundary-case handling
            return (-1.)

        if (start_query == end_query): #Issue/Boundary-case handling
            return res_scl
            
        ####################
        #Whole process/flow
        record_trav_dict[start_query] = True #Update/Record

        for [tmp_end_query, tmp_scl] in record_dict[start_query]:

            ##### Check if the current query matched conditions or not #####
            if (tmp_end_query not in record_trav_dict):
                res = self.dfsbfsTraversal(record_dict, tmp_end_query, end_query, (res_scl * tmp_scl), record_trav_dict) #Recursion function call
                if (res != (-1.)):
                    return res

        return (-1.)

File: Array/399.Evaluate_Division/C++/399.Evaluate_Division.cpp
class Solution {
public:
    vector<double> calcEquation(vector<vector<string>> & equations, vector<double> & values, vector<vector<string>> & queries) 
    {
        //====================================================//
        // Input type:                                        //
        //   - vector<vector<string>> ref. (equations vector) //
        //   - vector<double> ref. (values vector)            //
        //   - vector<vector<string>> ref. (queries vector)   //
        // Return type:                                       //
        //   - vector<double> (result vector)                 //
        //====================================================//

        /*Initialize*/
        ///// Length of equations, queries /////
        int len_equal = equations.size(), len_queries = queries.size();

        ///// Record, Traversal hashmap, dictionary /////
        unordered_map<string, vector<pair<string, double>>> record_map;
        unordered_map<string, bool> record_trav_map;

        ///// Record vector /////
        vector<string> record_vect;

        ///// Result vector /////
        vector<double> res_vect;


        /*DFS & BFS based loop traversal with recorded hashmap, dictionary*/
        ///// Step 1: Record indexed-string with hashmap, dictionary /////
        for (int equal_idx = 0; (equal_idx < len_equal); (equal_idx++)) //Whole
        {
            record_vect = equations[equal_idx]; //Record vector

            (record_map[record_vect[0]]).push_back(make_pair(record_vect[1], values[equal_idx])); //Keep updating, recording
            (record_map[record_vect[1]]).push_back(make_pair(record_vect[0], (1 / values[equal_idx]))); //Keep updating, recording

        } //Whole

        ///// Step 2: Looped-traversal with recorded hashmap /////
        for (int queries_idx = 0; (queries_idx < len_queries); (queries_idx++)) //Whole
        {
            record_vect = queries[queries_idx]; //Record vector

            res_vect.emplace_back(dfsbfsTraversal(record_map, record_vect[0], record_vect[1], record_trav_map, 1.0)); //Recursion function call

            record_trav_map.clear(); //Reset

        } //Whole

        return res_vect;
    }


    double dfsbfsTraversal(unordered_map<string, vector<pair<string, double>>> & record_map, string start_query, string & end_query, unordered_map<string, bool> & record_trav_map, double res_scl)
    {
        //===========================================================================================//
        // Input type:                                                                               //
        //   - unordered_map<string, vector<pair<string, double>>> ref. (record hashmap, dictionary) //
        //   - string (start query-string)                                                           //
        //   - string ref. (end query-string)                                                        //
        //   - unordered_map<string, bool> ref. (record traversal hashmap, dictionary)               //
        //   - double (result scaling-factor)                                                        //
        // Return type:                                                                              //
        //   - double (result scaling-factor))                                                       //
        //===========================================================================================//
        if ((record_map.find(start_query) ==  record_map.end()) || ((record_map.find(end_query) ==  record_map.end()))) //Issue, Boundary-case handling
        {
            return (-1.0);
        }

        if (start_query == end_query) //Issue, Boundary-case handling
        {
            return res_scl;
        }

        /*Initialize*/
        ///// Result scaling-factor /////
        double res;

        /*Whole process, flow*/
        record_trav_map[start_query] = true; //Update, Record

        for (pair<string, double> sub_pair: record_map[start_query]) //Whole
        {
            ///// Check if the current indexed-element existed or not /////
            if (record_trav_map.find(sub_pair.first) == record_trav_map.end())
            {
                res = dfsbfsTraversal(record_map, sub_pair.first, end_query, record_trav_map, (res_scl * sub_pair.second)); //Recursion fucntion call

                ///// Check if the current scaling-factor matched conditions or not /////
                if (res != (-1.0)) { return res; }
            }

        } //Whole

        return (-1.0);
    }
};
